src/tools/compiletest/src/main.rs

```diff
@@ -1,6 +1,24 @@
 use std::{env, io::IsTerminal, sync::Arc};
 
-use compiletest::{common::Mode, log_config, parse_config, run_tests};
+use compiletest::{
+    common::{Config, Mode},
+    log_config, parse_config, run_tests,
+};
+
+// Warn if filters point to a non .rs file
+fn warn_invalid_files(config: &Config) {
+    for filter in &config.filters {
+        let path = config.src_base.join(filter);
+        if path.is_file() && !path.extension().is_some_and(|ext| ext == "rs") {
+            eprintln!();
+            eprintln!(
+                "WARNING: Running {} test on an invalid source file: {}",
+                config.mode,
+                path.display(),
+            );
+        }
```

Comment:
Problem: so unfortunately, checking that the provided paths to `./x test` ends with an `rs` extension is both (1) an **overapproximation** and (2) an **underapproximation**.

1. **Overapproximation**: for example, the `rustdoc-js-std` test suite has primarily `js`-based tests and you can specify e.g. `./x test tests/rustdoc-js-std/parser-hof.js` and it would run that particular test.
2. **Underapproximation**: note that a file being `.rs` does not guarantee it is a test file; it can be an *auxiliary* file (support file), so the check presently would also miss passing an auxiliary `.rs` file.

I understand that this is a warning and is based on heuristics, but I wonder if we could try to be better here (so we don't generate misleading warnings, which I think is worse than not emitting a warning at all in this case).

I think something like the following logic may work (I need to do some experimenting myself):

1. Check if the provided path to a file has the test-suite-specific expected extension (e.g. `rs` for `ui`, `js` for `rustdoc-js-std`), otherwise easy bail.
2. If (1) succeeds, then we (somehow) panic in `run_tests` if 0 tests are run despite the pattern (of a path to a file) successfully matching.

Reply:
Thank you for the suggestions. I agree, the current heuristic is too flaky.

Are there any other test-suite-specific extensions I should check for?

Reply:
I *think* `rustdoc-js-std` is the special case here, but let me get back to you after I toy with this a bit myself (I need to experiment with where it's more logical to put this check/warning).

Reply:
Hey! Sorry for taking a long time to get back to you. I took a closer look at compiletest and discovered something funnier:

https://github.com/rust-lang/rust/blob/8a63c84af5e7a201239e87d2175128907495b028/src/tools/compiletest/src/lib.rs#L733-L744

so compiletest assumes all tests have `.rs` extension. This makes sense for what we use *compiletest* for, but the outer bootstrap layer supports other test suites as well!

Reply:
As in, when you run `./x test <filter>`, *bootstrap* is the one responsible for orchestrating *which* test runner is invoked. In the case of `rustdoc-js-std`, it's not compiletest! Let me go dig in bootstrap's test support infra to see if I can find out where this is done.

Reply:
Okay I returned from reading the bootstrap relevant logic. It seems that `rustdoc-js-std` already detects non-`.js` `<filter>` in the corresponding bootstrap step. With a better understanding, I think what we have here is mostly reasonable, but let me take a closer look at the implementation now.

Reply:
![relationship between bootstrap test dispatcher and test runners including compiletest](https://github.com/user-attachments/assets/11cb17f2-6d8d-4849-9626-75aa63f448c3)